notebook/nbextensions.py

```diff
@@ -208,24 +183,157 @@ def install_nbextension(path, overwrite=False, symlink=False, user=False, prefix
+        dest = nbext['dest']
+        require = nbext['require']
+        log(src, dest, require)
+        install_nbextension(src, overwrite=overwrite, symlink=symlink,
+            user=user, sys_prefix=sys_prefix, prefix=prefix, nbextensions_dir=nbextensions_dir,
+            destination=dest, verbose=verbose, log=log
+            )
+
+
+
+def uninstall_nbextension(dest, require, overwrite=False, symlink=False,
+                        user=False, sys_prefix=False, prefix=None, nbextensions_dir=None, 
+                        verbose=1, log=None):
+    """Uninstall a Javascript extension of the notebook
+    
+    Removes staged files and/or directories in the nbextensions directory and 
+    removes the extension from the frontend config.
+    
+    Parameters
+    ----------
+    
+    dest : str
+        path to file, directory, zip or tarball archive, or URL to install
+        name the nbextension is installed to.  For example, if destination is 'foo', then
+        the source file will be installed to 'nbextensions/foo', regardless of the source name.
+        This cannot be specified if an archive is given as the source.
+    require : str
+        require.js path used to load the extension
+    user : bool [default: False]
+        Whether to install to the user's nbextensions directory.
+        Otherwise do a system-wide install (e.g. /usr/local/share/jupyter/nbextensions).
+    prefix : str [optional]
+        Specify install prefix, if it should differ from default (e.g. /usr/local).
+        Will install to ``<prefix>/share/jupyter/nbextensions``
+    nbextensions_dir : str [optional]
+        Specify absolute path of nbextensions directory explicitly.
+    verbose : int [default: 1]
+        Set verbosity level. The default is 1, where file actions are self.log.infoed.
+        set verbose=2 for more output, or verbose=0 for silence.
+    """
+    nbext = _get_nbextension_dir(user=user, sys_prefix=sys_prefix, prefix=prefix, nbextensions_dir=nbextensions_dir)
+    dest = cast_unicode_py2(dest)
+    full_dest = pjoin(nbext, dest)
+    if os.path.lexists(full_dest):
+        if verbose >= 1:
+            log("Removing: %s" % full_dest)
+        if os.path.isdir(full_dest) and not os.path.islink(full_dest):
+            shutil.rmtree(full_dest)
+        else:
+            os.remove(full_dest)
+    
+    # Look through all of the config sections making sure that the nbextension
+    # doesn't exist.
+    data = _read_config_data(user=user, sys_prefix=sys_prefix)
+    if 'NotebookApp' in data:
+        notebook_app = data['NotebookApp']
+        for section in ['nbextensions_common', 'nbextensions_notebook', 'nbextensions_tree', 'nbextensions_edit', 'nbextensions_terminal']:
+            if section in notebook_app:
+                section_data = notebook_app[section]
+                if require in section_data:
+                    section_data[require] = None
+    _write_config_data(data, user=user, sys_prefix=sys_prefix)
```

Comment:
I don't think you want to write `data` back. Since deletions are partial, you only need to call `_write` with the deletions themselves.

I still think that these should be writing to `nbconfig/[section]/` and *not* `jupyter_notebook_config`.

```python
data = _read...
to_delete = {}
for section in ...:
    _write